### ISSUE TITLE
HiveMQ client doesn't send ping request to broker if smartwatch screen is off (Android)

### ISSUE THREAD

**Issue by cd** [hivemq/hivemq-mqtt-client#486]
## Expected behavior

HiveMQ android client should keep the connection to the broker alive with ping requests.

## Actual behavior

HiveMQ android client doesn't send a ping request if the screen of my android smartwatch goes off. After the timeout is exceeded, the client will (re)connect to the broker.

## Interesting notes

* The MQTT client runs inside a foreground service, so it is very stable and won't be killed from the system.
* Any kind of battery savings is turned off (network savings included)
* My app is whitelisted from doze mode.
* I turned off doze mode complety
* The MQTT client reconnect itself without any problems. Then why can't it just send a ping request?
* If I increase the keep alive time, it will take longer for the client to disconnect. However, the client always reacts immediately to the topic that it has subscribed to. So the connection is fine. It just doesn't ping.

## To Reproduce

This is the test scenario:
1. Open Android activity / start foreground service.
2. Init the MQTT client
3. Subscribe a random topic
4. Keep screen of smartwatch active (if not, the ping-req-res part will not happen, see below)
5. Just look at the logs

Wireshark:
![wireshark](https://user-images.githubusercontent.com/9284194/118318205-54f30080-b4f9-11eb-8d56-03420dd62e98.PNG)

Mosquitto broker:
![broker](https://user-images.githubusercontent.com/9284194/118318231-620fef80-b4f9-11eb-855f-b73b7549b0a4.PNG)

Here is some code from the android service:
```java
    private void initMqttClient() {
        client = MqttClient.builder()
                .useMqttVersion3()
                .identifier("new-fossil")
                .serverHost("192.168.0.17")
                .serverPort(1883)
                .automaticReconnectWithDefaultConfig()
                .addDisconnectedListener(context -> {
                    Log.d("MainService", "MQTT Disconnected");
                    ConnectivityManager cm = (ConnectivityManager)getSystemService(Context.CONNECTIVITY_SERVICE);
                    NetworkInfo activeNetwork = cm.getActiveNetworkInfo();
                    boolean isConnected = activeNetwork != null && activeNetwork.isConnected();
                    Log.d("MainService", context.getCause().getMessage());
                    if (!isConnectedToWifi()) {
                        Log.d("MainService", "(WIFI down)");
                    } else {
                        Log.d("MainService", "(WIFI OK)");
                    }
                })
                .addConnectedListener(context -> {
                    Log.d("MainService", "MQTT Connected");
                })
                .buildAsync();

        // Connect to MQTT broker
        client.connect();

        // Test counter
        subscribeTopic("testline/mytest/counter", 1, 56045);
    }
```

Logcat:
```
2021-05-14 20:29:11.291 3409-3437/com.mycomp.myapp D/MainService: MQTT Connected
2021-05-14 20:35:16.134 3409-3437/com.mycomp.myapp D/MainService: MQTT Disconnected
2021-05-14 20:35:16.138 3409-3437/com.mycomp.myapp D/MainService: Server closed connection without DISCONNECT.
2021-05-14 20:35:16.141 3409-3437/com.mycomp.myapp D/MainService: (WIFI OK)
2021-05-14 20:35:18.466 3409-3437/com.mycomp.myapp D/MainService: MQTT Connected
```

## Details
- Affected HiveMQ MQTT Client version(s): `com.hivemq:hivemq-mqtt-client:1.2.2`
- Used JVM version: 
- Used OS (name and version): Wear OS 2.27 (Android 9, API Level 28, H MR2)
- Used MQTT version: 3.1.1
- Used MQTT broker (name and version): Own mosquitto broker. I also tried the online test broker test.mosquitto.org.

I've tried a LOT of different things:
* Changed network/router
* Changed android configurations
* Changed different MQTT settings
* ...

I've been there for several days now and just can't get any further. Please help.

Thanks!

**Comment by SgtSilvio** [comment]
Hi @cd 
This is very likely caused by the device not waking up the application, so that the PING can not be sent.
Does the reconnect happen when the screen is still off or only after you turn the screen on again?

**Comment by cd** [comment]
Thanks for your reply! The reconnect also happen when the screen is off.

A workaround for me is, to create my own ping via an empty publish message
from my device, triggered by a timer. Or to subscribe a "ping topic" from
an anonther client.

I wonder what the basic problem here is. Technically, with a manual ping
workaround, it works.

The ability to manually send a real ping would at least be a little cleaner.


Silvio Giebl ***@***.***> schrieb am Fr., 21. Mai 2021, 18:24:

**Comment by SgtSilvio** [comment]
The problem seems to be that the ping task that is scheduled on an executor (Netty eventloop) does not trigger a wakeup of the device.
Closing of the connection by the broker (because it did not receive a ping) seems to wakeup the device, so the reconnect works.

**Comment by SgtSilvio** [comment]
@cd I have 2 asks:
- Can you run a simple scheduled task in a 60s interval that prints `System.nanoTime()` (which is used by the ping logic) when the screen is off?
- Can you create a netty eventloop and use this to schedule the task above when the screen is off?

**Comment by SgtSilvio** [comment]
Why I am asking:
> {@link #uptimeMillis} is counted in milliseconds since the
system was booted.  This clock stops when the system enters deep
sleep (CPU off, display dark, device waiting for external input),
but is not affected by clock scaling, idle, or other power saving
mechanisms.  This is the basis for most interval timing
such as {@link Thread#sleep(long) Thread.sleep(millls)},
{@link Object#wait(long) Object.wait(millis)}, and
{@link System#nanoTime System.nanoTime()}

https://android.googlesource.com/platform/frameworks/base/+/4118012/core/java/android/os/SystemClock.java

**Comment by cd** [comment]
To your questions:
* What kind of sheduled task do you mean? Timer Task? Alarm Manager? Job Sheduler? ...?
* I'm not familiar with netty and I don't know how to implement this in my android app.

But I have more knowledge now:
* I wrote that the reconnect works well. Unfortunately, this is no longer the case as soon as I have no longer connected the smartwatch to my computer via ADB. Then the app no longer reconnects. It's crazy.
* I've tested another smartwatch: An Oppo Watch with Wear OS 2.27 (also API Level 28). If I'm connected to the ADB server, then the watch even sends a ping. No reconnect required. But if I disconnect the watch from ADB, it's dead. No Ping, no reconnect.
* My already mentioned workaround, to send a "publish ping" from my watch via a timer task doesn't work anymore. It seems that after a while the operating system has stopped this behavior more and more. If I switch the screen on at any time (no matter what kind of screen/app), all timers are fired at once.

It is to despair...

**Comment by cd** [comment]
After a lot of trial and error, I have more and more the feeling that any kind of short, exact polling (ping, reconnect strategy) is very difficult or even impossible under Android / Wear OS. And even if there is some kind of "hack", that does not mean that it will work on all other devices and in the future in the same way.

I have now rebuilt my app logic and rely more on network-driven actions, because these seem to work reliably. This is how I get on (for now). Let's see what the other long-term tests say ...

Nonetheless, of course, I hope that a solution will still be found here.

Thank you for the good support and efforts up to this point!

**Comment by powturns** [comment]
This is how this issue was resolved for paho: https://github.com/eclipse/paho.mqtt.java/pull/775 to get the clocks working properly.

This is only a partial solution though, because even if you use an alarm receiver to wake up the device to preform a ping, the os may coalesce the wakeups such that they occur less frequently than the requested keepalive.

**Comment by cd** [comment]
Update: I am happy to announce that the ping problem does not occur with Wear OS 3.0 / OneUI Watch (Galaxy Watch 4). It remains to be seen whether this will also work as desired in future implementations. Then the problem would resolve itself in the medium term.

**Comment by HiveMQSupport** [comment]
Tune in to our upcoming webinars –– What's New in HiveMQ 4.7 & Testing MQTT in the Cloud Using MQTT.fx®

Dear hivemq/hivemq-mqtt-client,

It’s our pleasure to invite you to two of our upcoming webinars this month:

- What's New in HiveMQ: Inside the Upcoming HiveMQ 4.7 Release (https://cJgkn04.na1.hubspotlinks.com/Btc/LX+113/cJgkn04/VWTVQB61WST_W40f9QR7TWMxNW4dVnpw4xc2m_N2ZRHzN5mVy7V3Zsc37CgKgWMBKC3ckpWDHN8xBzSVdzjrdW7Mf9RG8ZtBKFW2_31v03FNG7RW71DwcC6sjNS0W1cXdly99W1ZCW7C-_4D5RqG0sW8zP9nh2W9Q2GW8dyJGc8LnrtwW7qB-jm3nh4CrW4hPVkr2qtwhQW8HcLlr3vCJCrN5PZ1vgzr_K6W14Y0bT3db-7rN39qSMWC-ZSxW1_NzGG31Q2GZN1b3yrDY5DkxW6KSzBk49dQDrW3kDzD722Sx3KW8yTHKn1QLDjPN8w75TRrNHx1W2HSGMT1Ym3cJW8rLQmv7J-SmHVxJSsW16dCp_W1ZPJvJ5z_fDhN33XFQsZ_TnCW1T2NyM1Lp_zbW88CbHg9cKZ6wW2pRbMg6QZN2DW3PZJYy6YdSkkVvNXdv3T5J1kW72bPsZ9d1cdjW6JxmRm7T9XZdVCwr774gNzQP3qgL1 )
- MQTT.fx® on HiveMQ Cloud: Testing MQTT in the Cloud is Now Easier Than Ever (https://cJgkn04.na1.hubspotlinks.com/Btc/LX+113/cJgkn04/VWTVQB61WST_W40f9QR7TWMxNW4dVnpw4xc2m_N2ZRHzN5mVy7V3Zsc37CgP8WW5d7X6g6_4-rMW4slNds4JfSC-W7rpDGb3DJ6KkW5BSntd4jf882W5RrRzW243x-gW3XsD1s20dHSlW7gp8Jt8B2VCbW57KZks7ZkY0jW2jGMqg47wS0JW4q5SJ86vTJY6W1Qfr4g9d9L3LW3t9W4c9jjgHsN9jGc3Lg3-XlW6qfRBs79h6tpW64BTK852Llk4W1jJGFk6KLGYRMbzXklgd7NyW2nWjfX6Rb_8TW4w1vRx78NhtQW2z-sMk3v0JGZVRcTqB4cB3bgN1pGtK17lSj0W8y3Hgj3fv3wgW1Cg5tm54ZDFvN22m7MMfb440W6dJL2Z8WTPtnW70DYfk4Cv1wrW1fJYl_9189GnF1XzQ6rZK0wW41PrXB84_WR1W858GjT7P6kz7W3ZlwTf8cYNW2W6QSr0h8rlHj5W7qrrw53L8nBt3bX-1 )

Why You Should Attend:

>> What's New in HiveMQ: Inside the Upcoming HiveMQ 4.7 Release

In the webinar, Dominik Obermaier, CTO and co-founder of HiveMQ, and Georg Held, Engineering Manager at HiveMQ, will showcase the new features of HiveMQ 4.7 and discuss how customers can benefit from upgrading to the new release.

Date: Wednesday, September 15. 2021

Time: 5 pm CEST / 11 am ET / 8 am PT

Register Now
(https://cJgkn04.na1.hubspotlinks.com/Btc/LX+113/cJgkn04/VWTVQB61WST_W40f9QR7TWMxNW4dVnpw4xc2m_N2ZRHzN5mVy7V3Zsc37CgBjhW2wKZ_j1qJR6LVVTsVX1MxVdmW3xnLsS91WtrCW8z62R-5fCffVN6dsbkNlfX3zW3X3wT24pFsQ3W37bDYs3bLBPBVlmPDH2wgmt3W4x1Pq11P0x0QW4FM71p3q7F84W1j22083GC6qkVh2f-n1RYsNYW7YwCkt2ZQMYDW2JmW1_5bdsLvW1cjWtW3c3lFzW10ZkwD1ZZnmfW4NSQYV7gPPqVW4XfHh43pS5v4W87Fjl68Qp_3RW2-G5cr8gl-nFN5Vzb44-NLbkW4Tl3xL7lwj1wVV11Y92WYR30W1dMPTF88Qh8wW2p-rth5GpVR9W1H-txN6fyFczW6gfyXs3DnKXKW8QvQfm97K7NRW8Jbz4g5p_p-_W70Hjrv3zxgR_W84c8D83wF9qTW8MBGLw2Mt3-1W31MBr66F5nhcW1p9hfz4bfz1d386d1 )

>> MQTT.fx® on HiveMQ Cloud: Testing MQTT in the Cloud is Now Easier Than Ever

In the webinar, Magi Erber, Product Manager at HiveMQ, will show you how simple it is to set up HiveMQ Cloud, on AWS or Azure, and increase the number of MQTT connections according to the demands of your IoT use case. Further in the session, Jens Deters, Founder and CEO of Softblade, and Nikolas Blättermann, COO of Softblade, together will show how easy it is to test, simulate and debug your MQTT client connections as you scale your IoT project with the new MQTT.fx® for HiveMQ Cloud edition.

Date: Wednesday, September 29. 2021

Time: 5 pm CEST / 11 am ET / 8 am PT

Register Now
(https://cJgkn04.na1.hubspotlinks.com/Btc/LX+113/cJgkn04/VWTVQB61WST_W40f9QR7TWMxNW4dVnpw4xc2m_N2ZRHzN5mVy7V3Zsc37CgYBLW8qthlK4jhJFXW7FLWs27l-6FSW6_TyLr3qPKfRW84FZ_H76Jj6bW15_qfF4X0L8bW1vHl1193WD8pW3ZS9Y9884lcJV3fNRV5lV07HW4yC4B_1L_qVjW1tKdyX3TWyjxW7bQxD93YXx9zW3zMxZQ16rtGLW69yymp57MPTRW1-yZ6m5WgZ3YN6N4-prxK9SJW3ttvLM5nSSCqW5BDqJL4XN04xW2m24kh7SwMDVW7G61Bc5P411PW5ldWTV99ScPpW85JrKd8ttvVcVfrf5D5Wz_tdN7-xxNDm3qsyW7S3tjs9346xLN5D4GJ51yrCYN62j2Gkvn5BFW1kwtFB89rhlSN1zJssRX_Fp0W1tnpCY5Dqlv5V6x9zK4fcL1JN4lTkTM_bhB5W15SB5Z81qJnyW1HXWSK5tNN_FW7TWCzd8Cv40w396m1 )

We look forward to seeing you at both these sessions.

Regards,

Your HiveMQ Marketing Team

HiveMQ GmbH, Ergoldinger Straße 2a, Landshut, Germany

Manage preferences (https://hs-5867930.s.hubspotemail.net/hs/manage-preferences/unsubscribe?languagePreference=en&amp%3Bd=VnBf_W5pR2lTW2p76LK3-1rCSW45Qvcz1ZtLFMW3ChgmW4hKDMFW3z9bNj1X0zbFW1_kht23KbsvFW3R0gxQ3ZYZMkW1X5dCf45RlhdW3P28X24hCPvhW43T2PJ8-zKhjW17f7hv8Z9KCRW8_4Znx5yXvv1W5GhwJ06D3sqMV28VhM5knbRVMGKd_P20N6LW41J87f4DFFTXV9wjJK2ChgQL51yGkb1&amp%3Bv=3&_hsenc=p2ANqtz--KWoNUTFkLbvO_eU94aLkcVjaXvtrWWqLvzPRj7gMp0lUcmFq1iQ1I14GQhy528qvvuk4qvloVlFHheckwECQGYUJ2N8u9ltK5F1gVdeqZ-P5iI6JkPO3f7G-BY8ejLzdBGW50PJIh1-Y07PwAOKuo6s2IOQ&_hsmi=156959882 )

You are receiving this email because you filled out a form on our website

or visited us at our booth at a tradeshow.

Twitter (https://cJgkn04.na1.hubspotlinks.com/Btc/LX+113/cJgkn04/VWTVQB61WST_W40f9QR7TWMxNW4dVnpw4xc2m_N2ZRHCf3pl3rV1-WJV7CgQ7DW3Z8-1h8gMtrWW4Nw9hM8rm0lLW7jHT-93_V96XN8qjJy8Pf56YW2Hwk8P21-Qr2W624my064DG8tW6MjdCg92V0DVW2lG2xf8lgvxCW7yGsfY46KzBlW7LCZ7517M8y2W7WQt2w6y0RxmVzTMqv9ghq5pW5t9QjD6626qbW28sFg67wl1jBW4mqwms169DLfW7YhwGJ76T9bxW3gRJjk1mH4ZRW3lKCb86GHN3JW2FqNv14fn9LzW5b3M175RtjStN5t6xS32XYLTW7-L_Yf8YKh_GN3Q_GZKQ7P6PW81SRVl2fjHqcW4M21gM6YddBwW5VRRdq7LKH46W7s9Y0n6Qmz3_W1pR-M_6s9gND3lB71 )

LinkedIn (https://cJgkn04.na1.hubspotlinks.com/Btc/LX+113/cJgkn04/VWTVQB61WST_W40f9QR7TWMxNW4dVnpw4xc2m_N2ZRHCy3pl3LV1-WJV7CgWMpW1G30Jq2FMDBGW8JtBY38fDyCcVG-7jc7kG3rkN25DvLkcjmpjW8jFBds4JpYScW6xTb7G1xCK6nW6L86Bc564QMwW5MhTcR8YLf3sW4CjTkk5Sd3bBN5pCt_tXNgXQW8YFT992Y2t_DW2hw4Yb5PgYWYN7HMhw1MxCTQW9k0lhD81S095W7yH_Fb4ZJQXZW2TB2_d27XMNdW17_GyS4qdSVQW4Yxx9H5t4lqgW2L6KYD7FY5RGW1DYqny391lDXW2YnTY-5-1rYbV3F98848yp3WW8nM4m13lXDnFW8wZlrQ82q2bJW8B5pqb5ns0FpW2rKnqP2Z4_0xW1r4M1P1fTKjZW6y691n8mDfytW2tfw281LHVfwW3859dN1_qzH-31rD1 )

Facebook (https://cJgkn04.na1.hubspotlinks.com/Btc/LX+113/cJgkn04/VWTVQB61WST_W40f9QR7TWMxNW4dVnpw4xc2m_N2ZRHCf3pl3rV1-WJV7CgGqrW30rdKS8JhcCqW4Xnyf14_BJcrW4RM27K30tYLTW1b7lYv3HsVXsW6VYnmn97zRPbV2lsgB35GSBFW8bdyBs5nXfMSW2Cg4H87wR4pzW5X4_m_6m1Qt0W3WThc16m6qN-W4FKQHz3HlgB-W7gnp3v76h-CZW3SnbS07rgSj4W4nwMtW4nc3GtN3VkxJCJr75bN8KCqVNL9crwV44WGl53zL1CW20QJ8x4cY0XdN5wbRm8bdkq_W6RRkcY6_dL-tW94z4zG15M0FNW3sT5BH1kv1TtN8YCDgSbhcWhW1l4YVm4wVTZYW5_8lg88P-5d8W44rst47vFPx7W3rYpVf4LcbP5W6vRhCC6wVk2y3g2M1 )

YouTube (https://cJgkn04.na1.hubspotlinks.com/Btc/LX+113/cJgkn04/VWTVQB61WST_W40f9QR7TWMxNW4dVnpw4xc2m_N2ZRHzt5mVxSV3Zsc37CgXvtV3XHWJ3P8xmwW2Ygv4797MvQ_W145KHC2SkhpbW3s8WhN1VQNSdW353LDm5-fCDRW3FM0b-25_YGmW7vBsNt8Ck-tsW7-_N4-7bLXRgW9dj8mp7hPrh4N8kqjxxHymZTW3K770l22pS81W4yJX_k6Qs8J_W6-jJLr8Y8Tx9W7rdK0w4kp6zyW98bgV47ydlCzW1RLzqr4wzfQBW3DkJ387rF-JfW7v1Jdj894hR8VmDyTJ32F665W94XqJp7TJ9gLW8VlvbR8b5PzbW9bkgW-45q1gTW8237q_19B0f0W3Wyk632yBp9tN3C7wys6539bW4sxqcl1lBGdKN9lq-RpgXshgW6rKZG27hLJ7HW59gDwX5x4X1vN3pSy-YNVbCcW6zJks_7ZXwhrW4tRfmj23p7lM3m3J1 )

**Comment by pglombardo** [comment]
Hi all - since this issue hasn't had an update in a couple years, I'm going to close it out.  If anything remains, please feel free to file a new issue - we'd be happy to help out!
